### PR TITLE
16.0 misc fidu fixes vin

### PR DESCRIPTION
**Pull request by vin-odoo**
Misc fix and bugprovement following fiscalis feedbacks

Task id # 3138441

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/base_vat/models/res_partner.py [hunk 1]
```diff
@@ -214,6 +230,29 @@ def _run_vat_test(self, vat_number, default_country, partner_is_company=True):
...
         return check_result
 
+    @api.model
+    def _run_vies_test(self, vat_number, default_country):
+        """ Validate a VAT number using the VIES VAT validation. """
+        check_result = None
+
+        # First check with country code as prefix of the TIN
+        vat_country_code, vat_number_split = self._split_vat(vat_number)
+        vat_has_legit_country_code = self.env['res.country'].search([('code', '=', vat_country_code.upper())])
```
Comment (FlorianGilbert) [inline addons/base_vat/models/res_partner.py:240]: self.env.ref or limit=1?

File: odoo/addons/base/models/res_partner.py [hunk 2]
```diff
@@ -61,6 +62,31 @@ def _view_get_address(self, arch):
...
+                        current_field = address_node.find(f".//field[@name='{field}']")
+                        if previous_field is not None and current_field is not None:
+                            previous_field.addnext(current_field)
+                        concerned_fields -= {field}
+                    # Add the remaining fields in 'concerned_fields' at the end, after the others
+                    for field in concerned_fields:
+                        previous_field = current_field
+                        current_field = address_node.find(f".//field[@name='{field}']")
+                        if previous_field is not None and current_field is not None:
+                            previous_field.addnext(current_field)
```
Comment (FlorianGilbert) [inline odoo/addons/base/models/res_partner.py:88]: :see_no_evil:

**Review by FlorianGilbert — CHANGES_REQUESTED** [review]

**Review by FlorianGilbert — APPROVED** [review]
LGTM

File: odoo/addons/base/models/res_partner.py [hunk 3]
```diff
@@ -61,6 +62,31 @@ def _view_get_address(self, arch):
...
+
```
Comment (qdp-odoo) [inline odoo/addons/base/models/res_partner.py:89]: heu..

there must be a nicer way of doing it... using_display_address and adding just a text node, a JS widget or even the above postprocess_and_fields() using address_view_id (why would it still be 2 different things, anyway?).

I'm not willing to allow this in stable: we do it cleanly or we do it cleanly in master only, but we do it clean anyway.

**Review by qdp-odoo — CHANGES_REQUESTED** [review]

File: odoo/addons/base/models/res_partner.py (hunk 3 above)
Comment (qdp-odoo) [reply odoo/addons/base/models/res_partner.py:89]: hackish but it's an FP request :man_shrugging:

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py [hunk 4]
```diff
@@ -596,6 +599,43 @@ def _import_fill_invoice_line_form(self, journal, tree, invoice, invoice_line, q
...
+            else:
+                tax = self.env['account.tax'].search(domain + [('price_include', '=', True)], limit=1)
+
+            # If we found a tax, get the related line, then check the amount. Correct it if needed.
+            if tax:
+                tax_line = invoice.line_ids.filtered(lambda line: line.tax_line_id == tax)
+                if tax_line:
+                    sign = -1 if invoice.is_inbound(include_receipts=True) else 1
+                    tax_line_amount = abs(tax_line.balance)
+                    if abs(tax_total - tax_line_amount) <= 0.05:
```
[the remark this answers is not in the document]
Comment (chklop) [reply addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py]: @qdp-odoo It is not useful, it is mandatory, we do the same in the OCR. It's a tax rounding problem. As the invoice is generated by another software who maybe does the tax rounding differently from the database in Odoo, we need to "correct" the rounding when you import it in Odoo. We cannot let Odoo change the invoice total.

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py [hunk 5]
```diff
@@ -596,6 +599,36 @@ def _import_fill_invoice_line_form(self, journal, tree, invoice, invoice_line, q
                 tax_nodes += elem.findall('.//{*}TaxSubtotal/{*}Percent')
         return self._import_fill_invoice_line_taxes(journal, tax_nodes, invoice_line, inv_line_vals, logs)
 
+    def _correct_invoice_tax_amount(self, tree, invoice):
+        """ The tax total may have been modified for rounding purpose, if so we should use the imported tax and not
+         the computed one """
+        if not invoice.is_invoice(include_receipts=True):
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:605]: useless check

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py [hunk 6]
```diff
@@ -487,11 +487,28 @@ def _import_fill_invoice_form(self, journal, tree, invoice, qty_factor):
...
+            if invoice_date_due_node is not None and invoice_date_due_node.text:
+                invoice.invoice_date_due = invoice_date_due_node.text
+                break
+
         # ==== Reference ====
 
         ref_node = tree.find('./{*}ID')
         if ref_node is not None:
-            invoice.ref = ref_node.text
+            if invoice.is_sale_document() and invoice.quick_edit_mode:
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:508]:
```suggestion
            if invoice.is_sale_document(include_receipts=True) and invoice.quick_edit_mode:
```

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py [hunk 7]
```diff
@@ -596,6 +599,36 @@ def _import_fill_invoice_line_form(self, journal, tree, invoice, invoice_line, q
...
+            return
+
+        # For each tax in our tax total, get the amount as well as the total in the xml.
+        all_taxes = self.env['account.tax'].search([
+            ('company_id', '=', invoice.company_id.id),
+            ('amount_type', '=', 'percent'),
+            ('type_tax_use', '=', invoice.journal_id.type),
+        ])
+        for elem in tree.findall('.//{*}TaxTotal/{*}TaxSubtotal'):
+            percentage = elem.find('.//{*}TaxCategory/{*}Percent')
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:615]: You are parsing twice exactly the same data...
At this point, you should already have the total for each account.tax, no need to parse it again.
You should perform the update of tax lines using the same mechanism as the pen on vendor bill

File: addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py [hunk 8]
```diff
@@ -596,6 +599,36 @@ def _import_fill_invoice_line_form(self, journal, tree, invoice, invoice_line, q
...
         return self._import_fill_invoice_line_taxes(journal, tax_nodes, invoice_line, inv_line_vals, logs)
 
+    def _correct_invoice_tax_amount(self, tree, invoice):
+        """ The tax total may have been modified for rounding purpose, if so we should use the imported tax and not
+         the computed one """
+        if not invoice.is_invoice(include_receipts=True):
+            return
+
+        # For each tax in our tax total, get the amount as well as the total in the xml.
+        all_taxes = self.env['account.tax'].search([
```
Comment (smetl) [inline addons/account_edi_ubl_cii/models/account_edi_xml_ubl_20.py:609]: wtf is that? `invoice.line_ids.tax_line_id`...

**Review by smetl — CHANGES_REQUESTED** [review]

**Comment by d-fence** [comment]
Upgrade exception [#180](https://runbot.odoo.com/web/#id=180&view_type=form&model=runbot.upgrade.exception) added.

Waiting the forward-port of this pr, the exception will be applied on all builds, and create an inconsistent state for migrations.
**Please forward port this asap up to master without change.** Exception should be forward-ported before the end of the week.

If you need to apply any change before it reaches master, please notify runbot team.

Details:
```
view:base_vat.view_partner_base_vat_form
field:res.users.vies_failed_message
field:res.partner.vies_failed_message
```
cc @KangOl @nseinlet

**Comment by qdp-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**